File: fboss/platform/helpers/PlatformNameLib.cpp
// (c) Meta Platforms, Inc. and affiliates. Confidential and proprietary.

#include "fboss/platform/helpers/PlatformNameLib.h"

#include <filesystem>

#include <fb303/ServiceData.h>
#include <folly/String.h>
#include <folly/logging/xlog.h>

namespace {

// Some platforms do not have the standardized platform-names in dmidecode yet.
// For such platforms, we use a translation function to get the standardized
// platform-names.
std::string sanitizePlatformName(const std::string& platformNameFromBios) {
  std::string platformNameUpper(platformNameFromBios);
  std::transform(
      platformNameUpper.begin(),
      platformNameUpper.end(),
      platformNameUpper.begin(),
      ::toupper);

  if (platformNameUpper == "MINIPACK3" ||
      platformNameUpper == "MINIPACK3_MCB") {
    return "MONTBLANC";
  }

  if (platformNameUpper == "JANGA") {
    return "JANGA800BIC";
  }

  if (platformNameUpper == "TAHAN") {
    return "TAHAN800BC";
  }

  return platformNameUpper;
}

} // namespace

namespace facebook::fboss::platform::helpers {

PlatformNameLib::PlatformNameLib(
    const std::shared_ptr<PlatformUtils> platformUtils,
    const std::shared_ptr<PlatformFsUtils> platformFsUtils)
    : platformUtils_(platformUtils), platformFsUtils_(platformFsUtils) {}

std::string PlatformNameLib::getPlatformNameFromBios(bool writeToCache) const {
  XLOG(INFO) << "Getting platform name from bios using dmedicode ...";
  auto [exitStatus, standardOut] =
      platformUtils_->execCommand(dmidecodeCommand);
  if (exitStatus != 0) {
    XLOG(ERR) << "Failed to get platform name from bios: " << stdout;
    throw std::runtime_error("Failed to get platform name from bios");
  }
  standardOut = folly::trimWhitespace(standardOut).str();
  XLOG(INFO) << "Platform name inferred from bios: " << standardOut;
  auto result = sanitizePlatformName(standardOut);
  XLOG(INFO) << "Platform name mapped: " << result;
  if (writeToCache) {
    platformFsUtils_->writeStringToFile(result, kCachePath, true);
  }
  return result;
}

std::optional<std::string> PlatformNameLib::getPlatformName() const {
  auto nameFromCache = platformFsUtils_->getStringFileContent(kCachePath);
  if (nameFromCache.has_value()) {
    auto result = nameFromCache.value();
    XLOG(INFO) << "Platform name read from cache: " << result;
    fb303::fbData->setCounter(kPlatformNameBiosReadFailures, 0);
    return result;
  }
  try {
    auto nameFromBios = getPlatformNameFromBios();
    fb303::fbData->setCounter(kPlatformNameBiosReadFailures, 0);
    return nameFromBios;
  } catch (const std::exception& e) {
    fb303::fbData->setCounter(kPlatformNameBiosReadFailures, 1);
    return std::nullopt;
  }
}

} // namespace facebook::fboss::platform::helpers

File: fboss/util/fixmyfboss/platform_name_lib.py
# pyre-strict
"""
Python implementation of platform name utilities for FBOSS.
This module is referenced from and provides equivalent functionality to:
https://www.internalfb.com/code/fbsource/fbcode/fboss/platform/helpers/PlatformNameLib.cpp
"""

from typing import Optional

from fboss.util.fixmyfboss.utils import run_cmd


def sanitize_platform_name(platform_name_from_bios: str) -> str:
    """
    Sanitize platform name from BIOS.
    Some platforms do not have the standardized platform-names in dmidecode yet.
    For such platforms, we use a translation function to get the standardized
    platform-names.
    """
    platform_name_upper = platform_name_from_bios.upper()

    name_translations = {
        "MINIPACK3": "MONTBLANC",
        "MINIPACK3_MCB": "MONTBLANC",
        "JANGA": "JANGA800BIC",
        "TAHAN": "TAHAN800BC",
        "MERU800BIAB": "MERU800BIA",
        "WEDGE800BNHP": "WEDGE800BACT",
        "WEDGE800CNHP": "WEDGE800CACT",
    }

    return name_translations.get(platform_name_upper, platform_name_upper)


def get_string_file_content(path: str) -> Optional[str]:
    """
    Read string content from file. Returns None if file doesn't exist or can't be read.
    """
    try:
        with open(path) as f:
            return f.read().strip()
    except Exception:
        return None


def get_platform_name_from_bios() -> str:
    """
    Returns platform name from dmidecode. Platform names are UPPERCASE but
    otherwise should match config names. Known dmidecode values are mapped to
    config names.
    """

    dmidecode_command = "dmidecode -s system-product-name"

    standard_out = run_cmd(dmidecode_command).stdout
    if not standard_out:
        error_msg = "Failed to get platform name from bios"
        raise RuntimeError(error_msg)

    standard_out = standard_out.strip()

    return sanitize_platform_name(standard_out)


def get_platform_name() -> Optional[str]:
    """
    Returns platform name matching get_platform_name_from_bios, but returns None
    instead of throwing exceptions.
    """
    cache_path = "/var/facebook/fboss/platform_name"
    name_from_cache = get_string_file_content(cache_path)

    if name_from_cache is not None:
        return name_from_cache

    try:
        return get_platform_name_from_bios()
    except Exception:
        return None
